src/webui/src/static/json_util.ts

```diff
@@ -0,0 +1,177 @@
+ *  @param keyOrKeys  Array of keys for each object,
+ *      or a single string as the same key of all objects,
+ *      or `undefined` if they are not dict value.
+ *
+ *  @returns  Formatted string for each object, without trailing comma.
+ **/
+function batchFormat(objects: any[], indent: string, width: number, keyOrKeys?: string | string[]): string[] {
+    let keys: string[];  // dict key as prefix string
+    if (keyOrKeys === undefined) {
+        keys = objects.map(() => '');
+    } else if (typeof keyOrKeys === 'string') {
+        keys = objects.map(() => `"${keyOrKeys}": `);
+    } else {
+        keys = keyOrKeys.map(k => `"${k}": `);
+    }
+
+    // try to collapse all
+    const lines = objects.map((obj, i) => keys[i] + stringifySingleLine(obj));
+    if (lines.every(line => (line.length + indent.length < width))) {
+        return lines;
+    }
+
+    // null values don't affect hierarchy detection
+    const nonNull = objects.filter(obj => obj !== null);
+    if (nonNull.length === 0) {
+        return lines;
+    }
+
+    if (Array.isArray(nonNull[0])) {
+        // objects are arrays, format all items in one batch
+        const elements = batchFormat(concat(nonNull), indent + '    ', width);
+        const iter = elements[Symbol.iterator]();
+        return objects.map((obj, i) => {
+            if (obj === null) {
+                return keys[i] + 'null';
+            } else {
+                return keys[i] + createBlock(indent, '[]', obj.map(() => iter.next().value));
+            }
+        });
+    }
+
+    if (typeof nonNull[0] === 'object') {
+        // objects are dicts, format values in one batch if they have similar keys
+        const values = concat(nonNull.map(obj => Object.values(obj)));
+        const childrenKeys = concat(nonNull.map(obj => Object.keys(obj)));
+        if (detectBatch(values)) {
+            // these objects look like TypeScript style `Map` or `Record`, where the values have same "type"
+            const elements = batchFormat(values, indent + '    ', width, childrenKeys);
+            const iter = elements[Symbol.iterator]();
+            return objects.map((obj, i) => {
+                if (obj === null) {
+                    return keys[i] + 'null';
+                } else {
+                    return keys[i] + createBlock(indent, '{}', Object.keys(obj).map(() => iter.next().value));
+                }
+            });
+
+        } else {
+            // these objects look like class instances, so we will try to group their fields
+            const uniqueKeys = new Set(childrenKeys);
+            const iters = new Map();
+            for (let key of uniqueKeys) {
+                const fields = nonNull.map(obj => obj[key]).filter(v => v !== undefined);
+                let elements;
+                if (detectBatch(fields)) {  // look like same field of class instances
+                    elements = batchFormat(fields, indent + '    ', width, key);
```

Comment:
Can we expose the interface of indent width?

Reply:
As long as it's not configurable in web UI, I think we can make life easier.